src/message.rs

```diff
@@ -0,0 +1,207 @@
+extern crate zmq_sys;
+
+use libc::{size_t};
+
+use std::ffi;
+use std::fmt;
+use std::{mem, ptr, str, slice};
+use std::os::raw::c_void;
+use std::ops::{Deref, DerefMut};
+
+use super::errno_to_error;
+
+/// Holds a 0MQ message.
+///
+/// A message is a single frame, either received or created locally and then
+/// sent over the wire. Multipart messages are transmitted as multiple
+/// `Message`s.
+///
+/// In rust-zmq, you aren't required to create message objects if you use the
+/// convenience APIs provided (e.g. `Socket::recv_bytes()` or
+/// `Socket::send_str()`). However, using message objects can make multiple
+/// operations in a loop more efficient, since allocated memory can be reused.
+pub struct Message {
+    msg: zmq_sys::zmq_msg_t,
+}
+
+impl Drop for Message {
+    fn drop(&mut self) {
+        unsafe {
+            let rc = zmq_sys::zmq_msg_close(&mut self.msg);
+            assert_eq!(rc, 0);
+        }
+    }
+}
+
+impl fmt::Debug for Message {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.deref())
+    }
+}
+
+unsafe extern fn drop_msg_content_box(data: *mut c_void, _hint: *mut c_void) {
+    let _ = Box::from_raw(data as *mut u8);
+}
+
+impl Message {
+
+    unsafe fn alloc<F>(f: F) -> Message where F: FnOnce(&mut zmq_sys::zmq_msg_t) -> i32 {
+        let mut msg = zmq_sys::zmq_msg_t::default();
+        let rc = f(&mut msg);
+        if rc == -1 {
+            panic!(errno_to_error())
+        }
+        Message { msg: msg }
+    }
+
+    /// Create an empty `Message`.
+    pub fn new() -> Message {
+        unsafe {
+            Self::alloc(|msg| { zmq_sys::zmq_msg_init(msg) })
+        }
+    }
+
+    /// Create a `Message` preallocated with `len` uninitialized bytes.
+    pub unsafe fn with_capacity_unallocated(len: usize) -> Message {
+        Self::alloc(|msg| { zmq_sys::zmq_msg_init_size(msg, len as size_t) })
+    }
+
+    /// Create a `Message` with space for `len` bytes that are initialized to 0.
+    pub fn with_capacity(len: usize) -> Message {
+        unsafe {
+            let mut msg = Message::with_capacity_unallocated(len);
+            ptr::write_bytes(msg.as_mut_ptr(), 0, len);
+            msg
+        }
+    }
+
+    /// Create a `Message` from a `&[u8]`. This will copy `data` into the message.
+    ///
+    /// This is equivalent to using the `From<&[u8]>` trait.
+    pub fn from_slice(data: &[u8]) -> Message {
+        unsafe {
+            let mut msg = Message::with_capacity_unallocated(data.len());
+            ptr::copy_nonoverlapping(data.as_ptr(), msg.as_mut_ptr(), data.len());
+            msg
+        }
+    }
+
+    fn from_box(data: Box<[u8]>) -> Message {
```

Comment:
Why not expose this one as well? If you have only a box (which I admit is not that likely) you'd have to copy it into a Vec.